### PR TITLE
quantized module serialization through prepack function registration

### LINKED ISSUE #108399: Generalize weight prepacking during quantized model deserialization
## Motivation

PyTorch only can serialize&deserialize parameters in quantized model for official supported QEngine now. There's no chance to extend it to achieve this goal for out-of-tree QEngine based on current code structure.  `torch.save/torch.load` would fail with a ScriptModule on QuantizedXPU backend.

Serializing quantized module (e.g. `ConvPackedParamsBase, LinearPrepackParamsBase`) depends on specific `prepack&unpack` implementation with regard to each QEngine.  We want to propose a mechanism that allows out-of-tree QEngines register their `prepack` implementation in PyTorch to finish the overall serialization process.

## Current behaviour
 Current pickling method call `prepack` of different backend based on compilation marco and global context `QEngine`. For each `QEngine`, code of using prepacking function to deserialize weight need to be added in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp#L481.  Like, `PackedLinearWeightsQnnp::prepack`, `PackedLinearWeightsOnednn::prepack` `PackedLinearWeight::prepack`

 There's no chance to extend it to out-of-tree QEngines.
 
Also, it is inconvenient for official supported qengine. For any new qengine added, it is required to add the code for deserialization, and maintainers need to know the some details of pickling method.

## Plan

Considering the `prepack` function has same signature for each `QEgnine`, we suggest to generalize the process of calling prepacking function. The prepacking function could be found at runtime based on the global context `QEngine`. Following is our proposal:

A dictionary is maintained by the library, where the key for each entry is `QEngine` and the value is the pointer to prepacking function of such engine. Each engine only need register the prepack function to the dictionary. When deserializing weight file, the `__getstate__` method can use `QEngine` as key to find the proper pointer to prepacking function. 

For maintainers of each official supported `QEngine`, only registering the prepacking function is needed (through `register_prepack_fn`).  No need to add code in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp#L481 anymore.

For maintainers of out-of-tree `QEngine`, maintainers can register prepacking function within their library.  Then function pointer is inserted to the map after registration and be found at runtime. No modifications is needed in PyTorch.  (Maintainers still need add enum `QEngine` in PyTorch.）

quantsave0

 @ezyan @arthuryuan1987  @gujinghui 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ZhiweiYan-96**
Fixes #108399

# Motivation

PyTorch only can serialize&deserialize parameters in quantized model for official supported QEngine now. There's no chance to extend it to achieve this goal for out-of-tree QEngine based on current code structure.  `torch.save/torch.load` would fail with a ScriptModule on QuantizedXPU backend.

Serializing quantized module (e.g. `ConvPackedParamsBase, LinearPrepackParamsBase`) depends on specific `prepack&unpack` implementation with regard to each QEngine.  We want to propose a mechanism that allows out-of-tree QEngines register their `prepack` implementation in PyTorch to finish the overall serialization process.

# Modification 
1. Add `register_prepack` `get_device_prepack_fn` utils pair for `qconv`, `qlinear` to register and query the pointer to prepacking function of each QEngine.
2. Add qengine `QXPU`, which is the engine for quantized XPU

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by ZhiweiYan-96** [comment]
@arthuryuan1987 @gujinghui

**Comment by gujinghui** [comment]
@ezyang could you help review? Thanks a lot.

**Comment by ezyang** [comment]
I'm not really qualified to review this. cc @jerryzh168 @vkuzo @supriyar

File: aten/src/ATen/native/quantized/QTensor.cpp [hunk 1]
```diff
@@ -177,7 +177,11 @@ Tensor& set_storage_quantized_(
   auto* self_ = self.unsafeGetTensorImpl();
   self_->set_storage_keep_dtype(std::move(storage));
   self_->set_storage_offset(storage_offset);
-  self_->set_sizes_and_strides(sizes, strides);
+  if(strides.data() == nullptr){
+    self_->set_sizes_contiguous(sizes);
+  }else{
```
Comment (jgong5) [inline aten/src/ATen/native/quantized/QTensor.cpp:182]: best you align the coding style with other code.
```suggestion
  } else {
```

File: aten/src/ATen/native/quantized/cpu/conv_serialization.h [hunk 2]
```diff
@@ -337,86 +337,38 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> deserialize_conv(
...
-    );
-  } // x86
-#endif
-
-#ifdef USE_FBGEMM
-  if (ctx.qEngine() == at::QEngine::FBGEMM) {
-    return PackedConvWeight<kSpatialDim>::prepack(
-      weight.value(),
+    if(use_onednn){
+      return get_device_prepack_fn<kSpatialDim>(at::QEngine::ONEDNN)(weight.value(),
```
Comment (jgong5) [inline aten/src/ATen/native/quantized/cpu/conv_serialization.h:346]: You removed `#ifdef USE_FBGEMM`. Then, this might be invoked even when `USE_FBGEMM` is not defined. Is it expected?

**Review by jgong5 — CHANGES_REQUESTED** [review]
A lot of failing tests...

File: aten/src/ATen/native/quantized/QTensor.cpp [hunk 3]
```diff
@@ -177,7 +177,11 @@ Tensor& set_storage_quantized_(
...
+    self_->set_sizes_and_strides(sizes, strides);
```
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/QTensor.cpp:183]: why is this change needed?

File: aten/src/ATen/native/quantized/cpu/conv_serialization.h (hunk 2 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/conv_serialization.h:346]: This code is also not meant to be compiled when not using FBGEMM

File: aten/src/ATen/native/quantized/cpu/conv_serialization.h [hunk 4]
```diff
@@ -337,86 +337,38 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> deserialize_conv(
...
-    );
-  }
-#endif // USE_PYTORCH_QNNPACK
-#if AT_MKLDNN_ENABLED()
-  if (ctx.qEngine() == at::QEngine::ONEDNN) {
-    return PackedConvWeightsOnednn<kSpatialDim>::prepack(
-      weight.value(),
+      transpose);
+  }else{
+    return get_device_prepack_fn<kSpatialDim>(eng)(weight.value(),
```
Comment (kimishpatel) [inline aten/src/ATen/native/quantized/cpu/conv_serialization.h:365]: if you are relying on `eng` then why specialize for fbgemm/onednn?

File: aten/src/ATen/native/quantized/cpu/conv_serialization.h (hunk 2 above)
Comment (ZhiweiYan-96) [reply aten/src/ATen/native/quantized/cpu/conv_serialization.h:346]: Hi @jgong5  @kimishpatel, Great thanks for the comments . The if-else branch here is intended to handle special engine `x86`, all other engine is handled by general query utils `get_device_preapck_fn`.  

By the way, any query for a unregistered/un-compiled engine prepack function would trigger TOCH_CHECK error.  So, it is within my expectation that these codes is compiled even no FBGEMM, due to the existence of prepack function is always checked.

Original logic in deserialize_conv can be simplified as follows. The x86 is a special engine due to bot onednn/fbgemm has chances to be used.
  ```c++
    #ifdef USE_FBGEMM
      if(engine == at::QEngine::x86){
        use_onednn = se_onednn = onednn_utils::should_use_onednn_quant(....);
        if(use_onednn){
          // call onednn prepack
        }else{
          // call fbgemm prepack
        }
      }
    #endif

    #ifdef USE_FBGEMM
      // handle fbgemm
    #endif

    #ifdef USE_QNNPACK
      // handle qnnpack
    #endif

    #if AT_MKLDNN_ENABLED
      // handle onednn
    #endif

    // fallback path
    // raise error
    TORCH_CHECK("Did not find xxxx")
  ```
  
  I intened to keep the above behavior. So it handles `x86` engine in `if branch`, all other engine is handled through 
  utils `get_device_prepack_fn`. If the prepack function is not compiled for given engine, then an error would be raised.
  ```c++
  if(engine == at::QEngine::x86){
    use_onednn = onednn_utils::should_use_onednn_quant(....);
    if(use_onednn){
      //call onednn prepack
    }else{
      // call fbgemm prepack, if not build with FBGEMM, RAISE ERROR
    }
  }else{
    // use engine to query prepack function, can handle qnnpack, onednn, fbgemm
    // IF NO PREPACK FUNCTION FOUND, RAISE ERROR
    prepack_fn = get_device_prepack_fn(engine)
    prepack_fn(...);
  }
  ```

File: aten/src/ATen/native/quantized/QTensor.cpp (hunk 3 above)
Comment (ZhiweiYan-96) [reply aten/src/ATen/native/quantized/QTensor.cpp:183]: hi @kimishpatel ,This modification is for the functionality of pickling non-cpu QTensor. 
  
Refer to https://github.com/pytorch/pytorch/pull/108400/files#diff-01ceffff32cac1db5af88e894b595f6e969d2db30b62f4e10842f336a7ca01e3L729, non-cpu Tensor is saved through putting it on host firstly. (`at::empty({0}, tensor.options()).set_(tensor.storage(), 0, size, {1}).cpu()`). 

 We add `at::empty_quantized({0}, tensor.options()).set_(tensor.storage(), 0, size, {1}).cpu()` for QTensor. 

`set_` would pass an empty stride for `set_storage_quantized`(https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/TensorShape.cpp#L373). But, the `set_storage_quantized` cannot handle an empty `strides`. This line modification is for QTensor to handle the case with empty strides (by treat it as an contiguous tensor).

File: aten/src/ATen/native/quantized/QTensor.cpp (hunk 1 above)
Comment (ZhiweiYan-96) [reply aten/src/ATen/native/quantized/QTensor.cpp]: great thanks for the suggestion, has modified

File: aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp [hunk 5]
```diff
@@ -568,11 +542,145 @@ int register_embedding_params() {
...
+DEFINE_PREPACK_FUNC_BY_DIM(3)
+
 namespace {
 
 static C10_UNUSED auto conv2d_params = register_conv_params<2>();
 static C10_UNUSED auto conv3d_params = register_conv_params<3>();
 static C10_UNUSED auto linear_params = register_linear_params();
 static C10_UNUSED auto embedding_params = register_embedding_params();
 
+static auto sel_2d = init_conv2d_prepack();
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp]: great thanks, has modified

File: aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp [hunk 6]
```diff
@@ -568,11 +542,145 @@ int register_embedding_params() {
...
+
+#ifdef USE_PYTORCH_QNNPACK
+  register_prepack<3>(at::QEngine::QNNPACK, (PackedConvWeightsQnnp<3>::prepack));
+#endif
+
+  return 0;
+}
+
+int init_linear_prepack(){
+#ifdef USE_PYTORCH_QNNPACK
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp]: thanks for the reminding, has fixed

**Comment by jerryzh168** [comment]
I heard from @jgong5 that this is the short term solution before you figure out how to move to 2.0, if this is just for CUDA, we have done similar things before, e.g.: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cudnn/LinearPrepack.cpp could you follow the same path? quantized engines are needed only when there are multiple implementations on the same device, in this case looks like you are trying to implement a different device, so I feel quantized engine is not needed

**Comment by gujinghui** [comment]
> I heard from @jgong5 that this is the short term solution before you figure out how to move to 2.0, if this is just for CUDA, we have done similar things before, e.g.: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cudnn/LinearPrepack.cpp could you follow the same path? quantized engines are needed only when there are multiple implementations on the same device, in this case looks like you are trying to implement a different device, so I feel quantized engine is not needed

Thanks @jerryzh168 

@ZhiweiYan-96 Let's evaluate and follow Jerry's suggestion.

**Comment by ZhiweiYan-96** [comment]
hi, @jerryzh168 great thanks for you helpful suggestion. 

**Quite agree that `device` is more suitable than `engine` here.**   We do have a prepack function(`PackedConvWeightQDPCPP`) registered in QuantizedXPU backend, just like cuda. We still need a chance for `deserialize_conv` to call non-CPU prepack function. 

Considering your advice, we suggest modify the current PR to **replacing qenige**(FBGEMM, ONEDNN) **with backend**(QuantizedCPU, QuantizedXPU, QuantizedCuda) **as the dictionary key**, and the value is the function registered in dispatching table for each backend. Here is a simple illustration. With following design, we have no need to add QXPU engine. All device impl is covered. For QuantizedCPU, we can directly call the code defined in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp. `deserialize_conv` could be simpler.
save_load_dict_backend_prepack_fn
)

May I know whether this more reasonable to you? Thanks~

I firstly want to re-implement whole `deserialize_conv` through operator dispatching. But seems there is no way in C++ code to call prepack like `torch.ops.quantized.conv2d_prepack(weight, bias)` in python.

**Comment by jerryzh168** [comment]
> hi, @jerryzh168 great thanks for you helpful suggestion.
> 
> **Quite agree that `device` is more suitable than `engine` here.** We do have a prepack function(`PackedConvWeightQDPCPP`) registered in QuantizedXPU backend, just like cuda. We still need a chance for `deserialize_conv` to call non-CPU prepack function.
> 
> Considering your advice, we suggest modify the current PR to **replacing qenige**(FBGEMM, ONEDNN) **with backend**(QuantizedCPU, QuantizedXPU, QuantizedCuda) **as the dictionary key**, and the value is the function registered in dispatching table for each backend. Here is a simple illustration. With following design, we have no need to add QXPU engine. All device impl is covered. For QuantizedCPU, we can directly call the code defined in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp. `deserialize_conv` could be simpler.
> 
> save_load_dict_backend_prepack_fn
> 
> May I know whether this more reasonable to you? Thanks~
> 
> I firstly want to re-implement whole `deserialize_conv` through operator dispatching. But seems there is no way in C++ code to call prepack like `torch.ops.quantized.conv2d_prepack(weight, bias)` in python.

I don't know why an additional dispatch (get_prepack_fn) is needed here, can you use something like https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cudnn/ConvPrepack.cpp#L206 to register your own prepack function?

**Comment by ZhiweiYan-96** [comment]
Hi @jerryzh168  The motivation we want to add additional dispatching is the `deserialize_conv` written in here only handle weight/bias for CPU device&engines(https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/conv_serialization.h#L287). And, `deserialize_conv` is accomplished through calling prepack function explicitly rather than operator dispatching.
All ConvPackedParams need use  `deserialize_conv` to load weight. I think both cuda/XPU cannot reuse this code for loading convweights?(https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp#L411).

**Comment by jerryzh168** [comment]
> Hi @jerryzh168 The motivation we want to add additional dispatching is the `deserialize_conv` written in here only handle weight/bias for CPU device&engines(https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/conv_serialization.h#L287). And, `deserialize_conv` is accomplished through calling prepack function explicitly rather than operator dispatching. All ConvPackedParams need use `deserialize_conv` to load weight. I think both cuda/XPU cannot reuse this code for loading convweights?(https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp#L411).

sorry still don't follow yet, can you implement a new class like PackedConvWeightsXPU: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/conv_serialization.h#L351?

**Comment by ZhiweiYan-96** [comment]
> sorry still don't follow yet, can you implement a new class like PackedConvWeightsXPU: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/conv_serialization.h#L351?

hi, @jerryzh168  Do you mean we can add a `PackedConvWeightsXPU` class directly defined in PyTorch like cuda, and directly call `PackedConvWeightsXPU::prepack` in `deserialize_conv`?  If that is ok to you, it's also easier for us, actually 👍

**Comment by jerryzh168** [comment]
> > sorry still don't follow yet, can you implement a new class like PackedConvWeightsXPU: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/conv_serialization.h#L351?
> 
> hi, @jerryzh168 Do you mean we can add a `PackedConvWeightsXPU` class directly defined in PyTorch like cuda, and directly call `PackedConvWeightsXPU::prepack` in `deserialize_conv`? If this is ok to you, it's also easier for us, actually

yeah that's what I meant, and the call should be guarded by some compilation flags similar to other backends I think

**Comment by ZhiweiYan-96** [comment]
```
collecting ... terminate called after throwing an instance of 'c10::Error'
  what():  Type c10::intrusive_ptr<ConvPackedParamsBase<2>, c10::detail::intrusive_target_default_null_type<ConvPackedParamsBase<2> > > could not be converted to any of the known types.
Exception raised from operator() at /home/gta/zhiwei/repos-baseline/pytorch/aten/src/ATen/core/jit_type.h:1811 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0xae (0x7ff59c2d4c8e in /home/gta/miniconda3/envs/zhiwei-baseline/lib/python3.10/site-packages/torch/
lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xf3 (0x7ff59c28995d in /home/gta/miniconda3/envs/zhiwei-baseline
/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #2: <unknown function> + 0x19e67ca (0x7ff59dd0c7ca in /home/gta/miniconda3/envs/zhiwei-baseline/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
frame #3: <unknown function> + 0x13b855d (0x7ff59d6de55d in /home/gta/miniconda3/envs/zhiwei-baseline/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
frame #4: c10::detail::infer_schema::make_function_schema(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&, c10::Arra
yRef<c10::detail::infer_schema::ArgumentDef>, c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>) + 0x60 (0x7ff59d6de9f0 in /home/gta/miniconda3/envs/zhiwei-baseline/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
frame #5: c10::detail::infer_schema::make_function_schema(c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>, c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>) + 0x6e (0x7ff59d6df92e in /home/gta/miniconda3/envs/zhiwei-baseli
ne/lib/python3.10/site-packages/torch/lib/libtorch_cpu.so)
```
@jerryzh168 I move our definition of prepack class into torch library, but we found above failures when try to `import torch`. Could you please have a look at the callstack? This issue stuck me for quite a time  :( . (I found a similar issue here https://github.com/pytorch/pytorch/issues/48452, but no idea what the root cause of this error)
